sympy/discrete/convolution.py

```diff
@@ -0,0 +1,194 @@
+"""
+Convolution (using FFT, NTT),
+Arithmetic Convolution (bitwise XOR, AND, OR), Subset Convolution
+"""
+from __future__ import print_function, division
+
+from sympy.core import S
+from sympy.core.compatibility import range, as_int
+from sympy.core.function import expand_mul
+from sympy.discrete.transforms import fft, ifft, ntt, intt
+
+
+def convolution(a, b, **hints):
+    """
+    Performs convolution by determining the type of desired
+    convolution using hints.
+
+    If no hints are given, linear convolution is performed using
+    FFT.
+
+    Parameters
+    ==========
+
+    a, b : iterables
+        The sequences for which convolution is performed.
+    hints : dict
+        Specifies the type of convolution to be performed.
+        The following hints can be given as keyword arguments.
+        dps : Integer
+            Specifies the number of decimal digits for precision for
+            performing FFT on the sequence.
+        prime : Integer
+            Prime modulus of the form (m*2**k + 1) to be used for
+            performing NTT on the sequence.
+        cycle : Integer
+            Specifies the length for doing cyclic convolution.
+
+    Examples
+    ========
+
+    >>> from sympy import S, I, convolution
+
+    >>> convolution([1 + 2*I, 4 + 3*I], [S(5)/4, 6], dps=3)
+    [1.25 + 2.5*I, 11.0 + 15.8*I, 24.0 + 18.0*I]
+
+    >>> convolution([1, 2, 3], [4, 5, 6], cycle=3)
+    [31, 31, 28]
+
+    >>> convolution([111, 777], [888, 444], prime=19*2**10 + 1)
+    [1283, 19351, 14219]
+
+    >>> convolution([111, 777], [888, 444], prime=19*2**10 + 1, cycle=2)
+    [15502, 19351]
+
+    """
+
+    dps = hints.pop('dps', None)
+    p = hints.pop('prime', None)
+    c = hints.pop('cycle', None)
```

Comment:
Here 0 could be the default instead of None

Reply:
Okay, so should `cycle=None` raise a `ValueError`?

Reply:
We can expect that if a user sets `cycle` then it would be a positive integer though 0 need not raise an exception.